doc.go

```diff
@@ -311,6 +326,14 @@ runes from the start of the line, and offset is a 0-based byte offset.
 The "text" field is the slice of bytes of the current match. It is empty
 in a predicate code block.
 
+The "state" field is a global store with backtrack support of type "map[string]interface{}",
```

Comment:
Should we add a comment explicitly noting that we use a "shallow" copy mechanism and thus pointers and structs containing pointer fields or similar (maps, slices) should be avoided? Another option would be to allow the client to specify a clone function for each of the keys that they set. What do you think?

Reply:
This is a very valid point. 

Just updating the documentation would be the easiest/cheapest way to handle this problem, but I am not satisfied with this solution.
I like the idea of a `cloner` interface, which could be implemented by the user to provide its own function, how the values inside the global state map should be cloned.

To better support the user, and to remove a source for failures, some kind of a DeepCopy functionality (maybe in combination with the `cloner` interface) would be the preferred solution in my opinion, even though I am not sure, if this is possible in every case, and for sure, this will not be a very fast solution.

So in the end, I would like to have a solution, which is save by default, but it should be possible to disable the save deepcopy, if more performance is needed by the user.

Reply:
As a first step, I updated this PR with an interface called `Cloner`, which allows the user to implement its own copy/clone method for the types used by him in the global state.
This is not yet the final solution, because it does not answer the question of the correct default behavior.

Possible solutions are:
* Just update the documentation, saying that the copying of the global state for backtracking purposes is only a "shallow" copy and that the user has to provide its own types and implement the Cloner interface, if he uses types, which are not properly cloned.
* Implement some kind of reflection based deep clone function within pigeon
* Use JSON marshal/unmarshal for the clone operation, which works for the most likely cases (map, slice), but will be most certainly quite slow.

It would also be possible to provide a "secure" default behavior (e.g. the JSON marshal variant) and let the user disable this functionality if he does not use such types or implemented the Cloner interface for all necessary cases (which would give him a performance gain).

Reply:
did some quick tests with JSON marshal/unmarshal. This is not a viable solution, because there is no guarantee to get the same types back. For example an `int` value will be changed to a `float64`.

Reply:
@jlozano Added the comment about the "shallow" copy for global state to the documentation.
I did a quick test with the mentioned deep copy library (https://github.com/mitchellh/copystructure). Even for a small example the parsing became three times slower. So I guess, it makes sense to push the decision how to clone (and what performance penalty is reasonable) to the user.

Do you mind to read the whole paragraph about global state in doc.go and give me feedback if it is understandable (and proper english).

Reply:
Suggested revisions:
```
The "state" field is a global store, with backtrack support, of type "map[string]interface{}".
The values in the store are tied to the parser's backtracking, in particular if a rule fails to match
then all updates to the state that occurred in the process of matching the rule are rolled back. 
For a key-value store that is not tied to the parser's backtracking, see the "globalStore".
The values in the "state" store are available for read access in action and predicate code blocks, any
changes made to the "state" store will be reverted once the action or predicate code block is
finished running. To update values in the "state" use state change code blocks ("#{}").
IMPORTANT:
- In order to properly roll back the state if a rule fails to match the parser must clone
the state before trying to match a rule.
- The default clone mechanism makes a "shallow" copy of each value in the "state", this implies
that pointers, maps, slices, channels, and structs containing any of the previous types are 
not properly copied.
- To support theses cases pigeon offers the "Cloner" interface which consists of a  single method 
"Clone". If a value stored in the "state" store implements this interface, the "Clone" method
is used to obtain a proper copy.
- If a general solution is needed, external libraries which provide deep copy
functionality may be user in the "Clone" method (e.g. https://github.com/mitchellh/copystructure).
```